### PR TITLE
[`distributed`] Add a new parallel plan for unit sharding

### PR DESCRIPTION
**Pull request by ArthurZucker**
[CPU CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48848) [GPU run-slow](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48848)

# What does this PR do?
This PR introduces unit sharding.
This affects mostly `q,k,v`, but i figured it will give more freedom in general. 

When a layer is expressed as 
```python
nn.Linear(hidden_dim, unit_number x unit_dim)
```
And you want to unsure the last dim is sharded to non empty `units`,  use this TP style.

`DTensor` does not allow uneven sharding. So the trick is to transform any `Shard(n)` to `Shard(n// units), Replicate()`.
This is not optimal in terms of memory, but its the only solution for now.

Superseeds #47811
Example:
```python
# 6 heads, 8 RANKs


```

### CODE REVIEW

File: src/transformers/distributed/tensor_parallel.py [hunk 1]
```diff
@@ -241,6 +243,79 @@ def transform_output_post_forward(self, module, output, mesh):
         return output.to_local() if self.use_local_output else output
 
 
+def _unit_mesh(mesh, unit_number: int):
```
Comment (3outeille) [inline src/transformers/distributed/tensor_parallel.py:246]: i think it's best to move this to `MeshManager` introduced in this PR https://github.com/huggingface/transformers/pull/48857 instead of carrying all the mesh

File: src/transformers/distributed/tensor_parallel.py [hunk 2]
```diff
@@ -241,6 +243,79 @@ def transform_output_post_forward(self, module, output, mesh):
...
+    return DeviceMesh(
+        device_type=mesh.device_type,
+        mesh=mesh.mesh.reshape(groups, replicas),
+        mesh_dim_names=[outer, "unit_replicate"],
+    )
+
+
+class UnitColwiseParallel(ColwiseParallel):
+    """This forces all ranks to own at least 1 unit of the tensor.
+    It is designed to be used on `query/key/value` layers when running TP. You cannot
```
Comment (3outeille) [inline src/transformers/distributed/tensor_parallel.py:263]:
```suggestion
    It is designed to be used on `query/key/value` layers when running TP. 
```

File: src/transformers/distributed/tensor_parallel.py [hunk 3]
```diff
@@ -241,6 +243,79 @@ def transform_output_post_forward(self, module, output, mesh):
...
+    For `query = nn.Linear(hidden_dim, head_dim * num_heads)` you can't divide the second dimension blindly by the world size as the attention REQUIRES to split per-head. You cannot split 2 heads of dim 64 into 4 GPUs by just separating into 4 x 32. This would mix the heads.
+
+    The unit colwise handles 2 cases:
+    - Less heads than world size: 2 heads, 4 process -> we replicate the heads.
+          Rank: | 0 |  1 |  2 | 3 |
+    This gives: |h0 | h0 | h1 | h1|
+
+    - More heads than world size: 6 heads, 8 process -> we also replicate the heads.
```
Comment (3outeille) [inline src/transformers/distributed/tensor_parallel.py:271]:
```suggestion
    - More heads than world size: 6 heads, 4 process -> we also replicate the heads.
```

File: src/transformers/distributed/tensor_parallel.py [hunk 4]
```diff
@@ -295,9 +370,6 @@ def context_around_forward(self, module, mesh):
         if not self.should_use_local_tensors(module):
             yield
         else:
-            # A rowwise local forward must produce only its partial matmul. If we don't hide
-            # the bias, we will be adding the bias x world_size times which is not correct.
```
Comment (3outeille) [inline src/transformers/distributed/tensor_parallel.py:299]: it is an important comment

File: tests/tensor_parallel/test_tensor_parallel.py [hunk 5]
```diff
@@ -365,3 +368,110 @@ def test_sharding_does_not_create_unrelated_module_attributes(self):
 
                 self.assertEqual(module.random_attr, 123)
                 self.assertFalse(hasattr(module, "num_experts"))
+
+
+@is_tensor_parallel_test
+class TestUnitParallel(TestCasePlus):
```
Comment (3outeille) [inline tests/tensor_parallel/test_tensor_parallel.py:374]: on top of that, we need forward + backward test in `test_tensor_parallel_mixin.py`. I am suspecting something might break training wise gradient wise

File: src/transformers/distributed/tensor_parallel.py [hunk 6]
```diff
@@ -813,6 +930,9 @@ def apply_tensor_parallelism(model, tp_mesh):
             style_name = _get_parameter_tp_plan(parameter_name=full, tp_plan=model.tp_plan, is_weight=True)
             if style_name is not None and style_name in ALL_PARALLEL_STYLES:
                 style = ALL_PARALLEL_STYLES[style_name]
+                if getattr(style, "needs_config", False):
```
Comment (3outeille) [inline src/transformers/distributed/tensor_parallel.py:933]: given we have to access the config as well when using `mla_kv_a_proj` style, i think it might better to just pass the config as actual argument now

File: src/transformers/distributed/tensor_parallel.py [hunk 7]
```diff
@@ -241,6 +243,79 @@ def transform_output_post_forward(self, module, output, mesh):
...
+        unit = weight.shape[0] // head_dim
+        unit_mesh = _unit_mesh(mesh, unit)
+        module._unit_mesh = unit_mesh
+        module._parameters[param] = torch.nn.Parameter(
+            distribute_tensor(meta, unit_mesh, [Shard(meta.ndim - 2), Replicate()], src_data_rank=None),
+            requires_grad=meta.requires_grad,
+        )
+
+    def transform_inputs_pre_forward(self, module, args, kwargs, mesh):
+        # The inputs are on a different mesh. We HAVE to swap the meshs as cross mesh coms
```
Comment (3outeille) [inline src/transformers/distributed/tensor_parallel.py:299]: nice

File: src/transformers/distributed/tensor_parallel.py [hunk 8]
```diff
@@ -241,6 +243,79 @@ def transform_output_post_forward(self, module, output, mesh):
...
+        # do not work.
+        # Swapping meshes is metadata only while the tensor is replicated, so gather first.
+        x = args[0]
+        if isinstance(x, DTensor):
+            replicated = [Replicate()] * x.device_mesh.ndim
+            if x.placements != tuple(replicated):
+                x = x.redistribute(placements=replicated)
+            x = x.to_local()
+        if not self.should_use_local_tensors(module):
+            x = DTensor.from_local(x, module._unit_mesh, [Replicate(), Replicate()], run_check=True)
```
Comment (3outeille) [inline src/transformers/distributed/tensor_parallel.py:309]: why `run_check=True` here ? in Colwise it is set to False to avoid extra comm

File: src/transformers/distributed/tensor_parallel.py [hunk 9]
```diff
@@ -334,6 +406,49 @@ def transform_output_post_forward(self, module, output, mesh):
...
+
+    def transform_inputs_pre_forward(self, module, args, kwargs, mesh):
+        x = args[0]
+        if self.should_use_local_tensors(module):
+            return (x.to_local() if isinstance(x, DTensor) else x,) + args[1:], kwargs
+        # `UnitColwiseParallel` already leaves its output on an equivalent unit mesh.
+        if not isinstance(x, DTensor):
+            x = DTensor.from_local(x, module._unit_mesh, [Shard(-1), Replicate()], run_check=True)
+        elif x.device_mesh != module._unit_mesh:
+            x = DTensor.from_local(x.to_local(), module._unit_mesh, [Shard(-1), Replicate()], run_check=True)
```
Comment (3outeille) [inline src/transformers/distributed/tensor_parallel.py:436]: same comment as for UnitColwise about why `run_check=True`

File: src/transformers/distributed/tensor_parallel.py [hunk 10]
```diff
@@ -334,6 +406,49 @@ def transform_output_post_forward(self, module, output, mesh):
...
+            requires_grad=meta.requires_grad,
+        )
+
+    def transform_inputs_pre_forward(self, module, args, kwargs, mesh):
+        x = args[0]
+        if self.should_use_local_tensors(module):
+            return (x.to_local() if isinstance(x, DTensor) else x,) + args[1:], kwargs
+        # `UnitColwiseParallel` already leaves its output on an equivalent unit mesh.
+        if not isinstance(x, DTensor):
+            x = DTensor.from_local(x, module._unit_mesh, [Shard(-1), Replicate()], run_check=True)
```
Comment (3outeille) [inline src/transformers/distributed/tensor_parallel.py:434]: same comment as for UnitColwise about why `run_check=True`

File: src/transformers/distributed/tensor_parallel.py [hunk 11]
```diff
@@ -334,6 +406,49 @@ def transform_output_post_forward(self, module, output, mesh):
...
+        elif x.device_mesh != module._unit_mesh:
+            x = DTensor.from_local(x.to_local(), module._unit_mesh, [Shard(-1), Replicate()], run_check=True)
+        return (x,) + args[1:], kwargs
+
+    def transform_output_post_forward(self, module, output, mesh):
```
Comment (3outeille) [inline src/transformers/distributed/tensor_parallel.py:439]: we should make it clear like RowiseParallel in `transform_output_post_forward` the 2 paths (inference & training)

File: src/transformers/distributed/tensor_parallel.py [hunk 12]
```diff
@@ -241,6 +243,79 @@ def transform_output_post_forward(self, module, output, mesh):
...
+                x = x.redistribute(placements=replicated)
+            x = x.to_local()
+        if not self.should_use_local_tensors(module):
+            x = DTensor.from_local(x, module._unit_mesh, [Replicate(), Replicate()], run_check=True)
+        return (x,) + args[1:], kwargs
+
+    def transform_output_post_forward(self, module, output, mesh):
+        # Already on the unit mesh, correctly laid out; the parent handles the local case.
+        if isinstance(output, DTensor):
+            return output
```
Comment (3outeille) [inline src/transformers/distributed/tensor_parallel.py:315]: this `if` will always return a dtensor  ? so potential mixed dtensor/tensor issue no ?

**Review by 3outeille — COMMENTED** [review]
Did a first review, we need some forward + backward test in `test_tensor_parallel_mixin.py` to be safe
